File: fuzzy_art.go
package art

import (
	"fmt"
	"math"
	"runtime"
	"slices"
	"sync"

	"art/internal/simd"
)

type fuzzyActivation struct {
	// fuzzy intersection
	fi []float64
	// L1 norm of the fuzzy intersection
	fiNorm float64
	// L1 norm of the relative category weights
	wNorm float64
	// activation value, choice function value
	activation float64
	// index of the category weights
	j int
}

type FuzzyART struct {
	workerPool chan struct{}
	batchSize  int
	wg         sync.WaitGroup

	// Vigilance parameter - controls category granularity
	// Recommended value: 0.86
	// Range: 0.0 to 1.0
	// Typical Values: 0.5 to 0.9
	// Purpose: Determines the strictness of category matching. Higher values mean stricter matching criteria.
	// Adjustment:
	// Increase rho to make the model more selective, creating more categories.
	// Decrease rho to allow more generalization, creating fewer categories.
	rho float64

	// Choice parameter - influences category competition
	// Recommended value: 0.01
	// Range: > 0.0
	// Typical Values: 0.0001 to 10
	// Purpose: Influences the activation function, affecting the competition among categories.
	// Adjustment:
	// Higher alpha values makes the choice more dependent on the match between input and weight
	// instead of considering the size of the input vector,
	// tend to favor the selection of existing categories over creating new ones.
	// Lower alpha values bias toward selecting categories with larger weight vectors.
	alpha float64

	// Learning rate - controls weight update speed
	// When beta == 1 is called fast-learning, it becomes sensitive to noise.
	// Lower β values provide more gradual, stable learning but require more training iterations.
	// Recommended value: 1.0
	// Range: 0.0 to 1.0
	// Purpose: Controls the rate at which the weights are updated during training.
	// Adjustment:
	// Increase beta for faster learning, which can be useful for rapidly changing environments.
	// Decrease beta for more stable learning, which can be beneficial for more stable environments.
	beta float64

	// M is the number of features of the input
	M int

	// W is the weight matrix - stores category prototypes
	W [][]float64

	// T is the activation list - stores category activations
	// This is an internal variable used for internal calculations
	// and should not be manipulated externally.
	T []*fuzzyActivation
}

func NewFuzzyART(inputLen int, rho float64, alpha float64, beta float64) (*FuzzyART, error) {
	if rho < 0 || rho > 1 {
		return nil, fmt.Errorf("vigilance parameter (rho) must be between 0 and 1, got %f", rho)
	}
	if alpha <= 0 {
		return nil, fmt.Errorf("choice parameter (alpha) must be positive, got %f", alpha)
	}
	if beta <= 0 || beta > 1 {
		return nil, fmt.Errorf("learning rate (beta) must be between 0 and 1, got %f", beta)
	}

	return &FuzzyART{
		workerPool: make(chan struct{}, runtime.NumCPU()),
		batchSize:  64,
		wg:         sync.WaitGroup{},
		rho:        rho,
		alpha:      alpha,
		beta:       beta,
		M:          inputLen,
		W:          make([][]float64, 0),
		T:          make([]*fuzzyActivation, 0),
	}, nil
}

// complementCode creates complement-coded representation of input vector.
// Complement coding is a common preprocessing step in ART models
// to prevent the "category proliferation problem."
// Complement coding achieve normalization while preserving amplitude information.
// Inputs preprocessed in complement coding are automatically normalized.
func (f *FuzzyART) complementCode(a []float64) []float64 {
	A := make([]float64, len(a)*2)
	for i, v := range a {
		A[i] = v
		A[i+len(a)] = 1 - v
	}

	return A
}

// activateCategories implements the recognition field functionality
// by computing activation values for each category based on the input vector.
// The sorting process also implicitly handles lateral inhibition by prioritizing
// the category with the highest activation, thereby inhibiting others.
func (f *FuzzyART) activateCategories(A []float64) {
	categoryChoice := func(input []float64, W [][]float64, startIndex int) {
		defer func() {
			// release the worker
			<-f.workerPool
			f.wg.Done()
		}()

		for i, w := range W {
			u := f.T[startIndex+i]
			u.j = startIndex + i
			u.fiNorm, u.wNorm = simd.Shared.FuzzyIntersectionNorm(A, w, u.fi)
			u.activation = u.fiNorm / (f.alpha + u.wNorm)
		}
	}

	for jStart := 0; jStart < len(f.W); jStart += f.batchSize {
		jEnd := jStart + f.batchSize
		if jEnd > len(f.W) {
			jEnd = len(f.W)
		}

		f.wg.Add(1)
		// acquire a worker
		f.workerPool <- struct{}{}

		// spawn a goroutine to process a batch of categories
		go categoryChoice(A, f.W[jStart:jEnd], jStart)
	}

	f.wg.Wait()
	f.sortCategoriesByActivation()
}

func (f *FuzzyART) sortCategoriesByActivation() {
	slices.SortFunc(f.T, func(a, b *fuzzyActivation) int {
		// In case of equal activation values, sort by category index,
		// because older categories must have the priority.
		if a.activation == b.activation {
			if a.j < b.j {
				return -1
			} else {
				return 1
			}
		}
		if a.activation > b.activation {
			return -1
		}
		return 1
	})
}

// resonance calculates the resonance between the input vector and a category.
// The resonance is the ratio of the fuzzy intersection L1 norm to the input vector L1 norm.
func (f *FuzzyART) resonance(fiNorm, aNorm float64) float64 {
	if fiNorm == 0 && aNorm == 0 {
		return 1
	}

	return fiNorm / aNorm
}

func (f *FuzzyART) appendNewCategory(A []float64) int {
	f.W = append(f.W, A)
	f.T = append(f.T, &fuzzyActivation{
		fi: make([]float64, len(f.W[0])),
	})
	return len(f.W) - 1
}

// resonateOrReset implements the resonance or reset logic.
// If the best matching category passes the vigilance test (>= rho),
// its weights are updated to move closer to the input vector, facilitating learning.
// If it fails, the category is inhibited (temporarily ignored),
// and the next best category is tested,
// continuing until a suitable category is found or all are exhausted
// in which case a new category is created.
func (f *FuzzyART) resonateOrReset(A []float64) (maxResonance float64, categoryIndex int) {
	aNorm := simd.Shared.SumFloat64(A)

	for _, t := range f.T {
		resonance := f.resonance(t.fiNorm, aNorm)
		if resonance >= f.rho {
			simd.Shared.UpdateFuzzyWeights(f.W[t.j], t.fi, f.beta)
			return resonance, t.j
		}
		maxResonance = math.Max(maxResonance, resonance)
	}

	// If no category meets the vigilance criterion, create a new category.
	// Fast commitment option, directly copy the input vector as the new category.
	categoryIndex = f.appendNewCategory(A)
	return
}

// Fit implements the complete ART learning cycle.
func (f *FuzzyART) Fit(a []float64) (resonance float64, categoryIndex int) {
	A := f.complementCode(a)
	f.activateCategories(A)
	return f.resonateOrReset(A)
}

// Predict implements the recognition process with optional learning.
// It returns the resonance value and the index of the best matching category.
// If learn is true, it also updates the weights of the matching category.
func (f *FuzzyART) Predict(a []float64, learn bool) (resonance float64, categoryIndex int) {
	A := f.complementCode(a)
	f.activateCategories(A)
	if !learn {
		aNorm := simd.Shared.SumFloat64(A)
		activation := f.T[0]
		resonance = f.resonance(activation.fiNorm, aNorm)
		return resonance, activation.j
	}

	return f.resonateOrReset(A)
}

func (f *FuzzyART) Close() {
	close(f.workerPool)
}

File: fuzzy_art.cpp
#include "fuzzy_art.hpp"
#include <iostream>
#include <numeric>
#include <mutex>
#include <cmath>

// Check if the CPU supports AVX-512
bool FuzzyART::checkAVX512Support() {
    #ifdef __AVX512F__
    return true;
    #else
    return false;
    #endif
}

FuzzyART::FuzzyART(size_t inputLen, float rho, float alpha, float beta) 
    : batchSize(16), rho(rho), alpha(alpha), beta(beta), inputDim(inputLen) {
    
    if (rho < 0 || rho > 1) {
        throw std::invalid_argument("Vigilance parameter (rho) must be between 0 and 1");
    }
    if (alpha <= 0) {
        throw std::invalid_argument("Choice parameter (alpha) must be positive");
    }
    if (beta <= 0 || beta > 1) {
        throw std::invalid_argument("Learning rate (beta) must be between 0 and 1");
    }
    
    // Check for AVX-512 support
    hasAVX512 = checkAVX512Support();
    
    // Initialize thread pool with hardware concurrency
    numThreads = std::thread::hardware_concurrency();
    pool = std::make_unique<ThreadPool>(numThreads);
    
    // Pre-allocate activation objects for the pool
    for (size_t i = 0; i < numThreads * batchSize; ++i) {
        activationPool.push_back(std::make_shared<Activation>(inputLen * 2));
    }
}

FuzzyART::~FuzzyART() {
    // Thread pool cleanup happens in its destructor
}

std::shared_ptr<Activation> FuzzyART::getActivation() {
    std::lock_guard<std::mutex> lock(poolMutex);
    if (activationPool.empty()) {
        // If pool is empty, create a new activation object
        return std::make_shared<Activation>(inputDim * 2);
    }
    
    auto activation = activationPool.back();
    activationPool.pop_back();
    return activation;
}

void FuzzyART::returnActivations(std::vector<std::shared_ptr<Activation>>& T) {
    std::lock_guard<std::mutex> lock(poolMutex);
    for (auto& t : T) {
        activationPool.push_back(t);
    }
    T.clear();
}

std::vector<float> FuzzyART::complementCode(const std::vector<float>& a) {
    std::vector<float> A(a.size() * 2);
    
    if (hasAVX512 && a.size() % 16 == 0) {
        // Use AVX-512 for complement coding if vector size is aligned
        __m512 ones = _mm512_set1_ps(1.0f);
        
        for (size_t i = 0; i < a.size(); i += 16) {
            // Load 16 elements from input vector
            __m512 x = _mm512_loadu_ps(&a[i]);
            
            // Store original values in first half
            _mm512_storeu_ps(&A[i], x);
            
            // Calculate and store complement (1-x) in second half
            __m512 complement = _mm512_sub_ps(ones, x);
            _mm512_storeu_ps(&A[i + a.size()], complement);
        }
    } else {
        // Fallback to scalar implementation
        for (size_t i = 0; i < a.size(); i++) {
            A[i] = a[i];                   // Original value
            A[i + a.size()] = 1.0f - a[i]; // Complement
        }
    }
    
    return A;
}

void FuzzyART::fuzzyIntersectionAVX512(const std::vector<float>& A, 
                                       const std::vector<float>& W, 
                                       std::vector<float>& result) {
    for (size_t i = 0; i < A.size(); i += 16) {
        // Load 16 elements from each vector
        __m512 a_vec = _mm512_loadu_ps(&A[i]);
        __m512 w_vec = _mm512_loadu_ps(&W[i]);
        
        // Compute element-wise minimum
        __m512 min_vec = _mm512_min_ps(a_vec, w_vec);
        
        // Store the result
        _mm512_storeu_ps(&result[i], min_vec);
    }
}

void FuzzyART::fuzzyIntersection(const std::vector<float>& A, 
                                const std::vector<float>& W, 
                                std::vector<float>& result) {
    if (hasAVX512 && A.size() % 16 == 0) {
        fuzzyIntersectionAVX512(A, W, result);
    } else {
        // Fallback to scalar implementation
        for (size_t i = 0; i < A.size(); i++) {
            result[i] = std::min(A[i], W[i]);
        }
    }
}

float FuzzyART::l1NormAVX512(const std::vector<float>& arr) {
    __m512 sum_vec = _mm512_setzero_ps();
    
    // Process 16 elements at a time
    for (size_t i = 0; i < arr.size(); i += 16) {
        __m512 vec = _mm512_loadu_ps(&arr[i]);
        sum_vec = _mm512_add_ps(sum_vec, vec);
    }
    
    // Horizontal sum of all elements in the vector
    return _mm512_reduce_add_ps(sum_vec);
}

float FuzzyART::l1Norm(const std::vector<float>& arr) {
    if (hasAVX512 && arr.size() % 16 == 0) {
        return l1NormAVX512(arr);
    } else {
        // Fallback to scalar implementation
        return std::accumulate(arr.begin(), arr.end(), 0.0f);
    }
}

std::pair<float, float> FuzzyART::categoryChoice(const std::vector<float>& A, 
                                                const std::vector<float>& W, 
                                                std::vector<float>& fuzzyIntersection) {
    fuzzyIntersection(A, W, fuzzyIntersection);
    float fiNorm = l1Norm(fuzzyIntersection);
    float choice = fiNorm / (alpha + l1Norm(W));
    return {choice, fiNorm};
}

std::vector<std::shared_ptr<Activation>> FuzzyART::activateCategories(const std::vector<float>& A) {
    std::vector<std::shared_ptr<Activation>> T(W.size());
    
    if (W.empty()) {
        return T;
    }
    
    // Synchronization for parallel processing
    std::mutex resultsMutex;
    std::condition_variable cv;
    size_t completed = 0;
    
    // Process categories in batches
    for (size_t jStart = 0; jStart < W.size(); jStart += batchSize) {
        size_t jEnd = std::min(jStart + batchSize, W.size());
        
        pool->enqueue([this, &A, &T, &resultsMutex, &cv, &completed, jStart, jEnd]() {
            std::vector<std::shared_ptr<Activation>> batchResults;
            
            for (size_t j = jStart; j < jEnd; ++j) {
                auto activation = getActivation();
                activation->j = j;
                
                // Compute category choice and fuzzy intersection
                auto [choice, fiNorm] = categoryChoice(A, W[j], activation->fi);
                activation->choice = choice;
                activation->fiNorm = fiNorm;
                
                batchResults.push_back(activation);
            }
            
            // Store results in the output vector
            {
                std::lock_guard<std::mutex> lock(resultsMutex);
                for (auto& act : batchResults) {
                    T[act->j] = act;
                }
                completed += batchResults.size();
            }
            
            cv.notify_one();
        });
    }
    
    // Wait for all batches to complete
    {
        std::unique_lock<std::mutex> lock(resultsMutex);
        cv.wait(lock, [&]{ return completed == W.size(); });
    }
    
    // Sort categories by activation values in descending order
    std::stable_sort(T.begin(), T.end(), 
        [](const std::shared_ptr<Activation>& a, const std::shared_ptr<Activation>& b) {
            // In case of equal activation values, sort by category index
            if (a->choice == b->choice) {
                return a->j < b->j;
            }
            return a->choice > b->choice;
        });
    
    return T;
}

float FuzzyART::matchCriterion(float fiNorm, float aNorm) {
    if (fiNorm == 0 && aNorm == 0) {
        return 1.0f;
    } else {
        return fiNorm / aNorm;
    }
}

std::pair<std::vector<float>, size_t> FuzzyART::resonateOrReset(
    const std::vector<float>& A,
    const std::vector<std::shared_ptr<Activation>>& T) {
    
    float aNorm = l1Norm(A);
    
    // Try to find a resonating category
    for (const auto& t : T) {
        if (!t) continue;  // Skip null activations
        
        float resonance = matchCriterion(t->fiNorm, aNorm);
        if (resonance >= rho) {
            // Update weights for the resonating category
            std::vector<float> newW(A.size());
            
            if (hasAVX512 && A.size() % 16 == 0) {
                __m512 beta_vec = _mm512_set1_ps(beta);
                __m512 one_minus_beta = _mm512_set1_ps(1.0f - beta);
                
                for (size_t k = 0; k < A.size(); k += 16) {
                    __m512 fi_vec = _mm512_loadu_ps(&t->fi[k]);
                    __m512 w_vec = _mm512_loadu_ps(&W[t->j][k]);
                    
                    // newW[k] = beta*t->fi[k] + (1-beta)*W[t->j][k]
                    __m512 beta_fi = _mm512_mul_ps(beta_vec, fi_vec);
                    __m512 old_w = _mm512_mul_ps(one_minus_beta, w_vec);
                    __m512 new_w = _mm512_add_ps(beta_fi, old_w);
                    
                    _mm512_storeu_ps(&newW[k], new_w);
                }
            } else {
                // Fallback to scalar implementation
                for (size_t k = 0; k < newW.size(); k++) {
                    newW[k] = beta * t->fi[k] + (1.0f - beta) * W[t->j][k];
                }
            }
            
            W[t->j] = newW;
            return {newW, t->j};
        }
    }
    
    // If no category meets the vigilance criterion, create a new category
    W.push_back(A);
    return {A, W.size() - 1};
}

std::pair<std::vector<float>, size_t> FuzzyART::fit(const std::vector<float>& a) {
    std::vector<float> A = complementCode(a);
    std::vector<std::shared_ptr<Activation>> T = activateCategories(A);
    auto result = resonateOrReset(A, T);
    returnActivations(T);
    return result;
}

std::pair<std::vector<float>, size_t> FuzzyART::predict(const std::vector<float>& a, bool learn) {
    std::vector<float> A = complementCode(a);
    std::vector<std::shared_ptr<Activation>> T = activateCategories(A);
    
    std::pair<std::vector<float>, size_t> result;
    if (!learn && !T.empty() && T[0]) {
        result = {W[T[0]->j], T[0]->j};
    } else {
        result = resonateOrReset(A, T);
    }
    
    returnActivations(T);
    return result;
}